Containerize course code execution (#1626)
PrairieLearn/PrairieLearn: docs/codeExecution.md

# Code Execution

> Central to PrairieLearn is the ability to securely execute arbitrary code. There are two primary contexts in which PrairieLearn must execute code:

Reviewer: ```suggestion
Central to PrairieLearn is the ability to securely execute arbitrary code. There are three primary contexts in which PrairieLearn must execute code:
```

* **Question and element code**: This is code from a question's `server.py` or any PrairieLearn/course [elements](elements.md). This code must execute as quickly as possible, as it will be executed any time a question is rendered or graded and thus needs to render during a single HTTP request.

> * [**Externally-graded questions**](externalGrading.md): This is code submitted by a student and graded for them by course code. This code can take longer to execute and is queued to be executed on distributed set of machines.

Reviewer: ```
* [**Workspaces**](workspaces/): This is an interactive environment in which a student can write and execute code, in contrast to the non-interactive batch execution by external graders. Workspaces can persist for many hours and so they are executed on a distributed set of machines (the _workspace hosts_, a distinct set from the _external grader hosts_).
```

> * [**Externally-graded questions**](externalGrading.md): This is code submitted by a student and graded for them by course code. This code can take longer to execute and is queued to be executed on distributed set of machines.

Reviewer: ```suggestion
* [**Externally-graded questions**](externalGrading.md): This is code submitted by a student and graded for them by course code. This code can take longer to execute and is queued to be executed on a distributed set of machines.
```

> Externally-graded questions and how they are executed is covered in more detail in the [external grading docs](externalGrading.md). This document is primarily concerned with describing how question and element code is executed.

Reviewer: ```suggestion
The [external grading docs](externalGrading.md) and [workspace docs](workspaces/) describe those execution modes in more detail. This document is primarily concerned with describing how question and element code is executed.
```

In production, we currently run PrairieLearn outside of Docker directly on an EC2 host.

To account for the variety of contexts in which PrairieLearn is executed, there are two related but distinct ways that PrairieLearn can execute question and element code: `native` and `container`. These modes correspond to the `workersExecutionMode` config value. They are described in more detail below.

### `native` execution mode

This is how PrairieLearn functioned historically. PrairieLearn would directly execute Python code with limited isolation from the rest of the system. This is largely the process described again, with a pool of zygotes.

This is still how PrairieLearn functions by default for local development. The `priairelearn/prairielearn` Docker image that is distributed to users includes all of the Python and R dependencies needed by question and element code, and said code is executed in the same container that PrairieLearn executes in. This is obviously bad for security, but doesn't matter for local development.

### `container` execution mode

> The `container` execution mode is currently used when PrairieLearn is running in production. It uses Docker to provide a degree of isolation from both PrairieLearn and other courses.

Reviewer 2: what does the 'and other courses' part mean here?

Reviewer 2: Oh, you mean so that the code from a course can't access files from PrairieLearn and from other courses?

## Code execution in practice

So far, this discussion has been pretty abstract. But what about all the actual code that underpins this stuff? Fear not, dear reader, we haven't forgotten about that!

### Code callers

A _code caller_ serves as an abstraction on top of the different execution modes above and hides the implementation details of exactly how code is executed. There are currently two different types of code callers, referred to here by the filenames of their implementations.

- `lib/code-caller-docker` handles executing code inside of Docker containers, as required by the `container` execution mode.
- `lib/code-caller-python` handles executing Python processes directly, as required by the `native` execution mode.

The primary external interface of these callers is the `call()` function, which takes five arguments:

> - `type`: the type of code being executed (either `question`, `course-element`, or `core-element`).

Reviewer 2: Is there a reason this distinction is important and these can't just all be executed the same way?

Author: Yes - the `type` determines a) where to resolve `directory` and b) whether or not to include `serverFilesCourse` on the Python interpreter's path.

- `directory`: the directory containing the file whose code will be executed.
  - For questions, this is an item in a course's `questions` directory.
  - For course elements, this is an item in a course's `elements` directory.
  - For core elements, this is an item in PrairieLearn's `elements` directory.

> - `file`: the name of the file whose code will be executed (e.g. `server.py`)

Reviewer 2: Why have a separate argument for directory and file? why not just pass the whole filepath as one argument?

Author: Technically, no, but doing it this way means that we isolate logic for resolving paths to one place. That is, we can just call `pc.call(type, directory, file)` without having to duplicate the logic for resolving a specific file from the combination of `(type, directory, file)` every time we invoke a python caller function.

2. That handler calls `getAndRenderVariant` in `lib/question` (a different function would be called if the user were submitting an answer).
3. That function calls an internal function that calls `render` in `question-servers/freeform.js`.
4. That function calls `getPythonCaller` in `lib/workers`. Depending on the active execution mode
   1. If running in `container` mode, a `lib/code-caller-docker` caller will be "prepared" for the current course (which sets up necessary bind mounts) and returned.
   2. If running in `native` mode, any available `lib/code-caller-python` caller will be returned.
5. `call(...)` is then repeatedly invoked on the code caller with the appropriate pieces of code to be executed.
6. Once the code caller is no longer needed during this request, `done()` is invoked on it. The forked worker is sent a `restart` message, which will cause the worker to exit and return control to the zygote. The zygote will then fork itself again, and the forked worker will wait until it receives more instructions.
7. Page render completes and the response is sent, thus finishing the request cycle.

## Operating in production

> When new versions of PrairieLearn are deployed, it's important to ensure that the appropriate executor image for the version being deployed is present on the machine. This ensures that PrairieLearn is able to serve traffic immediately instead of waiting for the new version to be pulled. If PrairieLearn is deployed with a `.git` directory present, the `./tools/script/pull.js` script can be used to log in to ECR and pull the appropriate version.

Reviewer 2: Maybe you want to define 'ECR' here? Unless everyone knows what it is already.

Author: This is referring to https://aws.amazon.com/ecr/, which is already a well-established part of how we handle Docker image s in production environments.

## The worker pool

A single PrairieLearn server may be serving potentially hundreds or thousands of assessments at one time. To handle this, we actually run a pool of zygotes described above that we call the _worker pool_. The pool maintains `N` zygotes and distributes requests to execute Python code across them. Requests are queued and handled in a FIFO basis. The worker pool also handles detecting unhealthy zygotes and replacing them with new ones.

## Execution modes

PrairieLearn must execute in two main environments: locally on the computers of people developing course content, and in a production environment. The primary differences between these two environments are a) ease of setup and b) whether or not PrairieLearn is running in a Docker container.

For local development, PrairieLearn must be easy to set up; it should not require complex infrastructure or commands to run. For this use case, PrairieLearn is distributed as a single Docker container that can be run without any external dependencies.

In production, we currently run PrairieLearn outside of Docker directly on an EC2 host.

> 

Author: Yeah, this "currently" description is going to be immediately out of date once this is deployed. I'll remove this.

> 

Author: Done in f98e288.

2. That handler calls `getAndRenderVariant` in `lib/question` (a different function would be called if the user were submitting an answer).
3. That function calls an internal function that calls `render` in `question-servers/freeform.js`.
4. That function calls `getPythonCaller` in `lib/workers`. Depending on the active execution mode
   1. If running in `container` mode, a `lib/code-caller-docker` caller will be "prepared" for the current course (which sets up necessary bind mounts) and returned.
   2. If running in `native` mode, any available `lib/code-caller-python` caller will be returned.
5. `call(...)` is then repeatedly invoked on the code caller with the appropriate pieces of code to be executed.
6. Once the code caller is no longer needed during this request, `done()` is invoked on it. The forked worker is sent a `restart` message, which will cause the worker to exit and return control to the zygote. The zygote will then fork itself again, and the forked worker will wait until it receives more instructions.
7. Page render completes and the response is sent, thus finishing the request cycle.

> ## Operating in production

Author: By that, do you mean self-hosted installations, or e.g. our own managed installations in CA/CN?

> ## Operating in production

Reviewer: Let's talk about this in person.

> ## Operating in production

Reviewer: Discussed.
